### PR TITLE
Some code change needed for Minpy examples

### PR DESCRIPTION
**Pull request by jermainewang**
Include:

1. ndarray functions for add/sub/mul/div
2. transpose
3. reduce/broadcasting support strictly in numpy semantics

### CODE REVIEW

File: python/mxnet/ndarray.py [hunk 1]
```diff
@@ -511,6 +641,99 @@ def ones(shape, ctx=None, dtype=mx_real_t):
...
+        reduction_shape = (before, between, after)
+        ret = ret.reshape(reduction_shape)
+        ret = sum_mid_internal(ret)
+        if len(shape) == 0:
+            ret = ret.reshape((1,)).asnumpy()[0]
+        else:
+            ret = ret.reshape(shape)
+    return ret
+
+def broadcast_to(a, shape):
```
Comment (tqchen) [inline python/mxnet/ndarray.py]: Add a dostring to this. Should it be a member function?

**Comment by tqchen** [comment]
- Please update the submodules to make sure submodules are up to date.
- Provide a few testcases to verify correctness of the new added ops

Thanks!

**Comment by jermainewang** [comment]
The scala-build cannot pass. It seems rely on the old `sum` function. Do you know who could have a look at this?

Also r-build failed in installing mxnet. Don't know how to fix them.

**Comment by tqchen** [comment]
I guess both could due to break change in libmxnet's sum function. Since they might be related to internal tests. 

Is it possible to keep the old function inact for now?

**Comment by jermainewang** [comment]
I know what happened, because the c++ `sum` function is removed due to the name duplication. I renamed it to be `sum_internal`. The problem is I need to use the `axis` and `keepdims` argument which is not available in the old sum function.

**Comment by tqchen** [comment]
cc @thirdwing @pluskid @javelinjs

**Comment by tqchen** [comment]
It is likely that we will need to add the wrappings of sum from all the current API wrappers. Let them do a pull request to your branch, to make things pass.

**Comment by jermainewang** [comment]
Another approach is to make a more general sum directly in c++ side.

**Comment by tqchen** [comment]
eventually things should go to c++ side. However, since we want to enable minpy support soon. I would recommend keep the current way.

Let us register the name by sum, but checks if the function already exist in current name space, if it exists, rename to function_internal

**Comment by tqchen** [comment]
some changes along this line might help  
https://github.com/dmlc/mxnet/blob/master/python/mxnet/ndarray.py#L834

**Pull request merged**